jobs/postgres-10/templates/drain.erb

```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+# If a command fails, in a pipeline or not, exit immediately
+set -e -o pipefail
+
+function prepend_datetime() {
+  awk -W interactive '{ system("echo -n [$(date +%FT%T%z)]"); print " " $0 }'
+}
+
+# Before redirecting stdout and stderr, copy the file descriptor for original
+# stdout where BOSH is expecting an integer, and only that.
+exec 3>&1
+exec \
+    1> >(prepend_datetime >> /var/vcap/sys/log/postgres-10/drain.out.log) \
+    2> >(prepend_datetime >> /var/vcap/sys/log/postgres-10/drain.err.log)
+
+output_for_bosh() {
+    exit_code=$?
+
+    if [ $exit_code -eq 0 ]; then
+        echo "postgres-10 deamon is properly shutdown with Fast Shutdown mode (SIGINT)"
+    else
+        echo "drain failed"
+    fi
+
+    echo $exit_code >&3
+}
+
+trap output_for_bosh EXIT
+
+
+postgres_pid=$(/var/vcap/packages/bpm/bin/bpm pid postgres-10)
+kill -s SIGINT "${postgres_pid}"
```

Comment:
This return immediately right? If so there is a change on a busy instance that bosh thinks draining is done and continues with sending a stop to bpm via monit.

An alternative would be to change to packaging script to also compile the pg_ctl script:
```
  pushd src/bin/pg_ctl
    make
    make install
  popd
```
And use the same call as the Postgres release: https://github.com/cloudfoundry/postgres-release/blob/develop/jobs/postgres/templates/postgres_ctl.sh.erb#L58

This should give more certainty that the process is actually stopping.

Reply:
No, we only need to fire-and-forget a SIGINT because BPM is currently not able to do it, and there is no need for any other complications.

With the drain script, here is the resulting sequence:
1. Monit is told not to restart Postgres if it stops.
2. the drain script sends a SIGINT and Postgres starts the fast shutdown. No need for wait loop or any timeout at this point. The drain script exits immediately. The point here is only to start the fast shutdown sequence of Postgres.
2. `monit stop` delegates to `bpm stop`. If the Postgres process is still alive, BPM sends a SIGTERM but Postgres is already doing a fast shutdown, so the signal is ignored.
3. BPM waits for 15 seconds and Postgres properly stop within this delay, whereas sending a SIGTERM only could have it hag for 5+ minutes.
4. If ever the 15 seconds delay expires, BPM would send a SIGQUIT and this would be perfectly legal because a fast shutdown is way shorter than that.

This explains why only sending a SIGINT from the drain script is enough.

Please also consider that this drain script is a fast-path solution. The better way to proceed is having BPM send a SIGINT directly, just like the Postgres release does. The PR is already pending, but this will take more time and we are in a hurry for our Postgres migration that has been blocked for 2+ weeks.

Reply:
I'm still confused about `The better way to proceed is having BPM send a SIGINT directly, just like the Postgres release does.` Since[ I could find no reference to SIGINT in the postgres-release](https://github.com/cloudfoundry/postgres-release/search?q=SIGINT). 

In the Postgres release they do ([source](https://github.com/cloudfoundry/postgres-release/blob/develop/jobs/postgres/templates/postgres_ctl.sh.erb#L58)):
```
su - vcap -c "${PACKAGE_DIR}/bin/pg_ctl stop -m fast -w -D ${DATA_DIR}"
```
```
-m
--mode=mode
  Specifies the shutdown mode. mode can be smart, fast, or immediate, or the first letter of one of these three. If this option is omitted, fast is the default.
```
```
-w
--wait
Wait for the operation to complete. This is supported for the modes start, stop, restart, promote, and register, and is the default for those modes.

When waiting, pg_ctl repeatedly checks the server's PID file, sleeping for a short amount of time between checks. Startup is considered complete when the PID file indicates that the server is ready to accept connections. Shutdown is considered complete when the server removes the PID file. pg_ctl returns an exit code based on the success of the startup or shutdown.

If the operation does not complete within the timeout (see option -t), then pg_ctl exits with a nonzero exit status. But note that the operation might continue in the background and eventually succeed.
```
source: https://www.postgresql.org/docs/10/app-pg-ctl.html

Reply:
No problem, it's only that `pg_ctl stop -m fast` sends a SIGINT signal to Postgres, as detailed the [Shutting Down the Server](https://www.postgresql.org/docs/10/server-shutdown.html) documentation.
> _The `pg_ctl` program provides a convenient interface for sending these signals to shut down the server_